yt/yt/library/containers/cri/image_cache.cpp

```diff
@@ -0,0 +1,482 @@
+                        image);
+                    cookie.Cancel(imageStatusOrError);
+                    imageStatusOrError.ThrowOnError();
+                }
+                auto& imageStatus = imageStatusOrError.Value();
+                if (!imageStatus->has_image()) {
+                    THROW_ERROR_EXCEPTION("Docker image is not found in cache after pull");
+                }
+                auto& protoImage = imageStatus->image();
+                auto imageId = TCriImageDescriptor{.Image = protoImage.id()};
+                YT_LOG_DEBUG("Docker image pull finished (Image: %v, ImageId: %v)",
+                    image,
+                    imageId);
+                auto imageFuture = cookie.GetValue();
+                DoInsertImage(cookie, protoImage, /*pullTime*/ TInstant::Now());
+                return imageFuture;
+            }));
+    }
+
+private:
+    bool IsManagedImage(const TCriImageDescriptor& image) const
+    {
+        for (const auto& prefix : Config_->UnmanagedPrefixes) {
+            if (image.Image.StartsWith(prefix)) {
+                return false;
+            }
+        }
+        if (Config_->ManagedPrefixes.empty()) {
+            return true;
+        }
+        for (const auto& prefix : Config_->ManagedPrefixes) {
+            if (image.Image.StartsWith(prefix)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    void DoInsertImage(TInsertCookie& cookie, const NCri::NProto::Image& protoImage, TInstant pullTime = TInstant::Zero())
+    {
+        TCriImageDescriptor imageName{.Image = cookie.GetKey()};
+        TCriImageDescriptor imageId{.Image = protoImage.id()};
+        i64 imageSize = protoImage.size();
+        bool pinned = false;
+
+        for (const auto& image : Config_->PinnedImages) {
+            if (image == imageName.Image) {
+                pinned = true;
+                break;
+            }
+        }
+
+        if (imageName.Image == imageId.Image) {
+            auto imageEntry = New<TCriImageCacheEntry>(imageId, imageId, imageSize);
+            cookie.EndInsert(std::move(imageEntry));
+        } else {
+            cookie.UpdateWeight(0);
+            auto imageCookie = BeginInsert(imageId.Image, imageSize);
+            if (imageCookie.IsActive()) {
+                auto imageEntry = New<TCriImageCacheEntry>(imageId, imageName, imageSize);
+                imageEntry->SetLastPullTime(pullTime);
+                imageCookie.EndInsert(std::move(imageEntry));
+            }
+            auto imageFuture = imageCookie.GetValue();
+            imageFuture.Subscribe(BIND([=, cookie = std::move(cookie)] (
+                        const TErrorOr<TCriImageCacheEntryPtr>& entryOrError) mutable
+            {
+                if (entryOrError.IsOK()) {
+                    auto imageEntry = entryOrError.Value();
+                    auto aliasEntry = New<TCriImageCacheEntry>(imageName, imageEntry);
```

Comment:
move

Reply:
nope

Reply:
Why?

Reply:
this is not subscribe-uniqe, we cannot deconstruct result here

Reply:
auto imageEntry = entryOrError.Value();

imageEntry was a copy, but I see that in the latest version there is a reference here